### PR TITLE
feat(hidden-moderator) Moderator can join hidden based on jwt

### PR DESCRIPTION
**Pull request by horymury**
- hidden moderator is not seen by other participants
- hidden moderator does not see their own thumbnail
- hidden moderator cannot unmute their video/audio or share screen
- hidden moderator can see and control other participants in meet( stop video, kick etc)

### CODE REVIEW

File: resources/prosody-plugins/mod_muc_moderator_hidden.lua [hunk 1]
```diff
@@ -0,0 +1,36 @@
+--This module adds is_hidden tag when the user from jwt is moderator and when
+--jwt contains `hidden` flag set to true, or removes it in case it was
+--added maliciously from client sent presence.
+--The module must be enabled under the muc component.
+local is_user_hidden = module:require "util".is_user_hidden;
+local tag_name = "is_hidden"
```
Comment (damencho) [inline resources/prosody-plugins/mod_muc_moderator_hidden.lua:6]: Isn't this the same as https://github.com/jitsi/jitsi-meet/blob/06bd279f225e2a6ea0c7121ca184e1c124a0d5f9/resources/prosody-plugins/mod_muc_poltergeist.lua#L146
Which we already handle in lib-jitsi-meet code?

File: resources/prosody-plugins/util.lib.lua [hunk 2]
```diff
@@ -239,6 +239,18 @@ function is_feature_allowed(session, feature)
     end
 end
 
+-- Utility function to check whether the user should be hidden.
+function is_user_hidden(session)
+    local context_user = session.jitsi_meet_context_user
+    if (context_user ~= nil
+        and (context_user['moderator'] == "true" or context_user['moderator'] == true)
```
Comment (damencho) [inline resources/prosody-plugins/util.lib.lua:246]: What is this context moderator? We do not have it anywhere in the code or documented?

File: react/features/follow-me/middleware.js [hunk 3]
```diff
@@ -104,11 +104,11 @@ function _onFollowMeCommand(attributes = {}, id, store) {
 
     // The Command(s) API will send us our own commands and we don't want
     // to act upon them.
-    if (participantSendingCommand.local) {
+    if (participantSendingCommand?.local) {
         return;
     }
 
-    if (participantSendingCommand.role !== 'moderator') {
+    if (participantSendingCommand && participantSendingCommand.role !== 'moderator') {
```
Comment (saghul) [inline react/features/follow-me/middleware.js:111]: Why this change? You should be able to use nullish coalescing now.

File: react/features/base/settings/functions.any.js [hunk 4]
```diff
@@ -265,6 +265,12 @@ export function shouldHideShareAudioHelper(state: Object): boolean {
  * @returns {boolean}
  */
 export function shouldHideSelfView(state: Object) {
+    const { conference } = state['features/base/conference'];
+
+    if (conference?.isHidden()) {
```
Comment (saghul) [inline react/features/base/settings/functions.any.js:270]: How do Jibris currently hide their local view?

File: react/features/participants-pane/functions.js [hunk 5]
```diff
@@ -235,6 +236,11 @@ export function getSortedParticipantIds(stateful: Object | Function): Array<stri
         dominant.push(dominantId);
     }
 
+    if (conference.isHidden()) {
```
Comment (saghul) [inline react/features/participants-pane/functions.js:239]: Do we really need to do this? We could disable the ability to raise the hand instead perhaps?

File: resources/prosody-plugins/mod_muc_moderator_hidden.lua (hunk 1 above)
Comment (saghul) [reply resources/prosody-plugins/mod_muc_moderator_hidden.lua:6]: Not sure. The target here is for a user to join but for the other users not see this user. This is for the "secretary manages the slides" scenario. Does poltergeist accomplish the same?

File: resources/prosody-plugins/util.lib.lua (hunk 2 above)
Comment (saghul) [reply resources/prosody-plugins/util.lib.lua:246]: This should be new. The idea is for this feature to only be available from the JWT since it's very sensitive.

File: react/features/follow-me/middleware.js (hunk 3 above)
Comment (horymury) [reply react/features/follow-me/middleware.js:111]: because here we're interested in the case in which participant is not hidden, and we check the role. If particpantSendingCommand is undefined, it does not mean it's not a moderator.

File: react/features/participants-pane/functions.js (hunk 5 above)
Comment (horymury) [reply react/features/participants-pane/functions.js:239]: this is in order to hide self in participants pane. Not related to raise hand

File: resources/prosody-plugins/util.lib.lua (hunk 2 above)
Comment (horymury) [reply resources/prosody-plugins/util.lib.lua:246]: `moderator` is a flag from JWT under user object. We only want to have the participant hidden if they are moderator.

File: react/features/base/settings/functions.any.js (hunk 4 above)
Comment (horymury) [reply react/features/base/settings/functions.any.js:270]: trough `config.iAmRecorder` and `config.iAmSipGateway`

File: resources/prosody-plugins/util.lib.lua (hunk 2 above)
Comment (damencho) [reply resources/prosody-plugins/util.lib.lua:246]: I don't like that we introduce new field in the jwt that is not mentioned/documented or used anywhere in this repo.
Maybe we need to add a module which respects that and makes a participant a moderator. As the similar module from jitsi-contrib.

File: resources/prosody-plugins/mod_muc_moderator_hidden.lua (hunk 1 above)
Comment (damencho) [reply resources/prosody-plugins/mod_muc_moderator_hidden.lua:6]: Maybe, I don't know what does this flag do. In the other case we ignore presence so we do not see a participant leave and join. I suppose both end up doing the same.

Comment (damencho) [reply resources/prosody-plugins/mod_muc_moderator_hidden.lua:6]: Ah you need the presence actually to be able to do certain operations.

File: resources/prosody-plugins/util.lib.lua (hunk 2 above)
Comment (horymury) [reply resources/prosody-plugins/util.lib.lua:246]: hmm I would not call `moderator` a new jwt field, I think it is used for quite a while now. Probably until now it was only used client-side?

Comment (damencho) [reply resources/prosody-plugins/util.lib.lua:246]: Maybe, I'm missing it, I do not see it anywhere in this repo? Or in the doc about the tokens: https://github.com/jitsi/lib-jitsi-meet/blob/master/doc/tokens.md

Comment (horymury) [reply resources/prosody-plugins/util.lib.lua]: removed the moderator check

**Comment by horymury** [comment]
Jenkins please test this please.

File: resources/prosody-plugins/mod_muc_user_hidden.lua [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+function add_hidden_tag(event)
+    local stanza = event.stanza;
+    local session = event.origin;
+
+    if stanza == nil or stanza.name ~= "presence" then
+      return
+    end
+
+    stanza:maptags(function(tag)
+      if tag and tag.name == tag_name then
```
Comment (damencho) [inline resources/prosody-plugins/mod_muc_user_hidden.lua:19]: Why do we need to do that on every presence? 
I think this will be problematic in big meetings as this will be executed on every presence that pass through prosody.

Comment (damencho) [reply resources/prosody-plugins/mod_muc_user_hidden.lua:19]: Answer to myself: to make sure nobody else is using it ... yeah this worries me that will have a high impact on prosody cpu usage, especially on setting up a big meeting.

Comment (damencho) [reply resources/prosody-plugins/mod_muc_user_hidden.lua:19]: Nope ignore the session part.

Maybe we can add a feature to the features advertised and the client activates all the logic here only if this is enabled, otherwise even with the tag in presence it will not work.

Comment (damencho) [reply resources/prosody-plugins/mod_muc_user_hidden.lua:19]: Even better that flag can be used only in lib-jitsi-meet. And if the lib is not adding the isHidden nothing from the logic works.

Comment (horymury) [reply resources/prosody-plugins/mod_muc_user_hidden.lua:19]: you mean adding the is_hidden on self presence? that would mean parsing the JWT client side...

**Comment by damencho** [comment]
After offline discussion closing for https://github.com/jitsi/jitsi-meet/pull/10973
